pkg/bloomfilter/filter.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package bloomfilter
+
+import (
+	"encoding/binary"
+	"math"
+	"math/rand"
+
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/pkg/storj"
+)
+
+const (
+	version1 = 1
+)
+
+// rangeOffsets contains offsets for selecting subranges
+// that minimize overlap in the first hash functions
+var rangeOffsets = [...]byte{9, 13, 19, 23}
+
+// Filter is a bloom filter implementation
+type Filter struct {
+	seed      byte
+	hashCount byte
+	table     []byte
+}
+
+// newExplicit returns a new custom filter.
+func newExplicit(seed, hashCount byte, sizeInBytes int) *Filter {
+	return &Filter{
+		seed:      seed,
+		hashCount: hashCount,
+		table:     make([]byte, sizeInBytes),
+	}
+}
+
+// NewOptimal returns a filter based on expected element count and false positive rate.
+func NewOptimal(expectedElements int, falsePositiveRate float64) *Filter {
+	seed := byte(rand.Intn(255))
+
+	// calculation based on https://en.wikipedia.org/wiki/Bloom_filter#Optimal_number_of_hash_functions
+	bitsPerElement := int(-1.44*math.Log2(falsePositiveRate)) + 1
+	hashCount := int(float64(bitsPerElement)*math.Log(2)) + 1
```

Comment:
Why do we have the `+ 1` in the above two formulas? I don't see this coming from the Wikipedia link.

Perhaps, you mentioned it yesterday, but I don't remember...

Reply:
@egonelbre ^

Reply:
Also the formula for hashCount does not seem correct. Shouldn't it be: **k = -log<sub>2</sub>p**

Reply:
Not sure. I guess we should separate this logic out and implement tests for the optimal params. @Fadila82

Reply:
For the +1, it's to be sure to have enough bits per element and hashes to achieve the desired probability. It's the same than ceil(float)

For the hashCount, it also says that k= (m/n) ln 2, which is what is used.

Reply:
Based on https://hur.st/bloomfilter

```
filterSizeInBits = ceil((elementCount * log(p)) / log(1 / pow(2, log(2))));
hashCount = round((filterSizeInBits / elementCount) * log(2));
```

The `+1` yeah looks wrong. But it might be to avoid zero condition.

Reply:
For `ceil(x)` is (almost) the same as `floor(x + 1)`... however here we should be ceiling to the next byte or larger, since we cannot create a bucket with less bits... i.e. `ceil(x/32)*32`. And we probably cannot remove the size for that same reason.

For the `round`, that would need to be `int((filterSize / elementCount) * math.Log(2) + 0.5)`.

Reply:
(I'm not sure how large the practical difference is.)

Reply:
https://play.golang.org/p/iWPAwX_uDIN

Reply:
and with rounding to nearest uint64 https://play.golang.org/p/feVrJ7HyWHK

Reply:
It seems to achieve better in terms of size of the bloom filter, but it gives a false positive rate higher than the targeted one. 
For instance, if I try it with an false positive rate of 0.1, the current one gives:
0.0964
0.0878
0.087
0.1018
0.0916

With the second one:
0.1274
0.124
0.1136
0.1164
0.119

And with the third:
0.1274
0.124
0.1136
0.1164
0.119

Reply:
I think you may want to try keeping it as a float all the way to `sizeInBytes` and only then rounding up.

```go
	// calculation based on https://en.wikipedia.org/wiki/Bloom_filter#Optimal_number_of_hash_functions
	bitsPerElement := -1.44 * math.Log2(falsePositiveRate)
	hashCount := bitsPerElement * math.Log(2)
	if hashCount > 255 {
		// it will never be larger, but just in case to avoid overflow
		hashCount = 255
	}
	sizeInBytes := int(math.Ceil(float64(expectedElements) * bitsPerElement / 8))

	return newExplicit(seed, byte(hashCount), sizeInBytes)
```

Reply:
I wouldn't have expected this. I guess we can leave it as is then.

But, this suggests there is some problem somewhere.

Reply:
Anyways, I reran the calculations:

Stats from different implementations, also uploaded the diagnostic tool:

```
mse = mean-squared-error from the expected false positive
hc  = hash count
sz  = size of the table
```

current:
```
n=    1000 p=0.01 avg=0.01 min=0.00 mean=0.01 max=0.02 mse=0.000 hc=7 sz=1.2 KiB
n=   10000 p=0.01 avg=0.01 min=0.00 mean=0.01 max=0.02 mse=0.000 hc=7 sz=12.2 KiB
n=  100000 p=0.01 avg=0.01 min=0.00 mean=0.01 max=0.02 mse=0.000 hc=7 sz=125.0 KB
n= 1000000 p=0.01 avg=0.01 min=0.00 mean=0.01 max=0.02 mse=0.000 hc=7 sz=1.2 MB
n=    1000 p=0.05 avg=0.03 min=0.02 mean=0.03 max=0.05 mse=0.000 hc=5 sz=0.9 KiB
n=   10000 p=0.05 avg=0.03 min=0.02 mean=0.04 max=0.05 mse=0.000 hc=5 sz=8.5 KiB
n=  100000 p=0.05 avg=0.03 min=0.02 mean=0.03 max=0.05 mse=0.000 hc=5 sz=85.4 KiB
n= 1000000 p=0.05 avg=0.03 min=0.02 mean=0.03 max=0.06 mse=0.000 hc=5 sz=0.9 MB
n=    1000 p=0.10 avg=0.09 min=0.07 mean=0.09 max=0.11 mse=0.000 hc=4 sz=625 B
n=   10000 p=0.10 avg=0.09 min=0.07 mean=0.09 max=0.11 mse=0.000 hc=4 sz=6.1 KiB
n=  100000 p=0.10 avg=0.09 min=0.07 mean=0.09 max=0.12 mse=0.000 hc=4 sz=61.0 KiB
n= 1000000 p=0.10 avg=0.09 min=0.07 mean=0.09 max=0.12 mse=0.000 hc=4 sz=625.0 KB
n=    1000 p=0.20 avg=0.15 min=0.12 mean=0.15 max=0.20 mse=0.003 hc=3 sz=500 B
n=   10000 p=0.20 avg=0.15 min=0.12 mean=0.15 max=0.18 mse=0.003 hc=3 sz=5.0 KB
n=  100000 p=0.20 avg=0.15 min=0.12 mean=0.15 max=0.17 mse=0.003 hc=3 sz=50.0 KB
n= 1000000 p=0.20 avg=0.15 min=0.11 mean=0.15 max=0.18 mse=0.003 hc=3 sz=500.0 KB
n=    1000 p=0.30 avg=0.25 min=0.22 mean=0.25 max=0.29 mse=0.002 hc=3 sz=375 B
n=   10000 p=0.30 avg=0.25 min=0.22 mean=0.25 max=0.28 mse=0.002 hc=3 sz=3.7 KiB
n=  100000 p=0.30 avg=0.25 min=0.22 mean=0.25 max=0.28 mse=0.002 hc=3 sz=36.6 KiB
n= 1000000 p=0.30 avg=0.25 min=0.22 mean=0.25 max=0.28 mse=0.003 hc=3 sz=375.0 KB
```

version 1:
```
n=    1000 p=0.01 avg=0.01 min=0.00 mean=0.01 max=0.02 mse=0.000 hc=10 sz=1.2 KiB
n=   10000 p=0.01 avg=0.01 min=0.01 mean=0.01 max=0.02 mse=0.000 hc=10 sz=11.7 KiB
n=  100000 p=0.01 avg=0.01 min=0.01 mean=0.01 max=0.02 mse=0.000 hc=10 sz=117.0 KiB
n= 1000000 p=0.01 avg=0.01 min=0.01 mean=0.01 max=0.02 mse=0.000 hc=10 sz=1.1 MiB
n=    1000 p=0.05 avg=0.05 min=0.04 mean=0.05 max=0.07 mse=0.000 hc=6 sz=0.8 KiB
n=   10000 p=0.05 avg=0.06 min=0.04 mean=0.05 max=0.08 mse=0.000 hc=6 sz=7.6 KiB
n=  100000 p=0.05 avg=0.06 min=0.04 mean=0.05 max=0.07 mse=0.000 hc=6 sz=76.1 KiB
n= 1000000 p=0.05 avg=0.06 min=0.04 mean=0.06 max=0.08 mse=0.000 hc=6 sz=0.7 MiB
n=    1000 p=0.10 avg=0.11 min=0.09 mean=0.11 max=0.14 mse=0.000 hc=5 sz=599 B
n=   10000 p=0.10 avg=0.11 min=0.09 mean=0.11 max=0.14 mse=0.000 hc=5 sz=5.8 KiB
n=  100000 p=0.10 avg=0.11 min=0.09 mean=0.11 max=0.14 mse=0.000 hc=5 sz=58.5 KiB
n= 1000000 p=0.10 avg=0.12 min=0.09 mean=0.12 max=0.15 mse=0.000 hc=5 sz=585.0 KiB
n=    1000 p=0.20 avg=0.21 min=0.18 mean=0.21 max=0.25 mse=0.000 hc=3 sz=418 B
n=   10000 p=0.20 avg=0.21 min=0.17 mean=0.21 max=0.24 mse=0.000 hc=3 sz=4.1 KiB
n=  100000 p=0.20 avg=0.21 min=0.18 mean=0.20 max=0.24 mse=0.000 hc=3 sz=40.9 KiB
n= 1000000 p=0.20 avg=0.21 min=0.17 mean=0.21 max=0.24 mse=0.000 hc=3 sz=408.9 KiB
n=    1000 p=0.30 avg=0.34 min=0.30 mean=0.34 max=0.39 mse=0.002 hc=3 sz=313 B
n=   10000 p=0.30 avg=0.34 min=0.29 mean=0.34 max=0.37 mse=0.002 hc=3 sz=3.1 KiB
n=  100000 p=0.30 avg=0.34 min=0.29 mean=0.34 max=0.39 mse=0.002 hc=3 sz=30.6 KiB
n= 1000000 p=0.30 avg=0.34 min=0.30 mean=0.34 max=0.38 mse=0.002 hc=3 sz=305.9 KiB
```

jeffs calculation:
```
n=    1000 p=0.01 avg=0.01 min=0.00 mean=0.01 max=0.02 mse=0.000 hc=7 sz=1.2 KiB
n=   10000 p=0.01 avg=0.01 min=0.00 mean=0.01 max=0.02 mse=0.000 hc=7 sz=11.7 KiB
n=  100000 p=0.01 avg=0.01 min=0.00 mean=0.01 max=0.02 mse=0.000 hc=7 sz=116.8 KiB
n= 1000000 p=0.01 avg=0.01 min=0.00 mean=0.01 max=0.02 mse=0.000 hc=7 sz=1.1 MiB
n=    1000 p=0.05 avg=0.05 min=0.03 mean=0.05 max=0.07 mse=0.000 hc=5 sz=0.8 KiB
n=   10000 p=0.05 avg=0.05 min=0.03 mean=0.05 max=0.07 mse=0.000 hc=5 sz=7.6 KiB
n=  100000 p=0.05 avg=0.05 min=0.04 mean=0.05 max=0.07 mse=0.000 hc=5 sz=76.0 KiB
n= 1000000 p=0.05 avg=0.05 min=0.04 mean=0.05 max=0.07 mse=0.000 hc=5 sz=0.7 MiB
n=    1000 p=0.10 avg=0.10 min=0.07 mean=0.10 max=0.13 mse=0.000 hc=4 sz=598 B
n=   10000 p=0.10 avg=0.10 min=0.08 mean=0.10 max=0.13 mse=0.000 hc=4 sz=5.8 KiB
n=  100000 p=0.10 avg=0.10 min=0.09 mean=0.10 max=0.13 mse=0.000 hc=4 sz=58.4 KiB
n= 1000000 p=0.10 avg=0.10 min=0.08 mean=0.10 max=0.13 mse=0.000 hc=4 sz=583.9 KiB
n=    1000 p=0.20 avg=0.21 min=0.17 mean=0.21 max=0.24 mse=0.000 hc=3 sz=418 B
n=   10000 p=0.20 avg=0.21 min=0.17 mean=0.21 max=0.25 mse=0.000 hc=3 sz=4.1 KiB
n=  100000 p=0.20 avg=0.21 min=0.18 mean=0.21 max=0.25 mse=0.000 hc=3 sz=40.8 KiB
n= 1000000 p=0.20 avg=0.21 min=0.17 mean=0.21 max=0.23 mse=0.000 hc=3 sz=408.2 KiB
n=    1000 p=0.30 avg=0.30 min=0.27 mean=0.30 max=0.35 mse=0.000 hc=2 sz=313 B
n=   10000 p=0.30 avg=0.30 min=0.27 mean=0.31 max=0.34 mse=0.000 hc=2 sz=3.1 KiB
n=  100000 p=0.30 avg=0.30 min=0.27 mean=0.30 max=0.34 mse=0.000 hc=2 sz=30.5 KiB
n= 1000000 p=0.30 avg=0.30 min=0.26 mean=0.30 max=0.34 mse=0.000 hc=2 sz=305.3 KiB
```

Reply:
Well, I still wonder if it's ok to have a fp rate higher than the targeted one. Are we ok with that?

Reply:
Yeah, it's fine.